connectors/cache/fallback.go

```diff
@@ -0,0 +1,306 @@
+package cache
+
+import (
+	"context"
+	"errors"
+	"sort"
+	"sync"
+
+	"github.com/uber-go/dosa"
+	"github.com/uber-go/dosa/connectors/base"
+	"github.com/uber-go/tally"
+)
+
+const (
+	key   = "key"
+	value = "value"
+)
+
+// Scope is a namespace wrapper around a stats reporter, ensuring that
+// all emitted values have a given prefix or set of tags.
+type Scope interface {
+	// Counter returns the Counter object corresponding to the name.
+	Counter(name string) tally.Counter
+
+	// Tagged returns a new child scope with the given tags and current tags.
+	Tagged(tags map[string]string) Scope
+
+	// SubScope returns a new child scope appending a further name prefix.
+	SubScope(name string) Scope
+}
+
+type rangeResults struct {
+	Rows      []map[string]dosa.FieldValue
+	TokenNext string
+}
+
+type rangeQuery struct {
+	Conditions map[string][]*dosa.Condition `json:",omitempty"`
+	Token      string
+	Limit      int
+}
+
+// NewConnector creates a fallback cache connector
+func NewConnector(origin dosa.Connector, fallback dosa.Connector, encoder Encoder, scope Scope, entities ...dosa.DomainObject) *Connector {
+	bc := base.Connector{Next: origin}
+	set := createCachedEntitiesSet(entities)
+	return &Connector{
+		Connector:         bc,
+		origin:            origin,
+		fallback:          fallback,
+		encoder:           encoder,
+		cacheableEntities: set,
+		stats:             scope,
+	}
+}
+
+// Connector is a fallback cache connector
+type Connector struct {
+	base.Connector
+	origin            dosa.Connector
+	fallback          dosa.Connector
+	encoder           Encoder
+	cacheableEntities map[string]bool
+	mux               sync.Mutex
+	stats             Scope
+	// Used primarily for testing so that nothing is called in a goroutine
+	synchronous bool
+}
+
+// SetCachedEntities sets the entities that will write and read from the fallback
+func (c *Connector) SetCachedEntities(entities ...dosa.DomainObject) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	c.cacheableEntities = createCachedEntitiesSet(entities)
+}
+
+// Upsert dual writes to the fallback cache and the origin
+func (c *Connector) Upsert(ctx context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
+	w := func() error {
+		cacheKey := createCacheKey(ei, values, c.encoder)
+		cacheValue, err := c.encoder.Encode(values)
+		if err != nil {
+			return err
+		}
+		adaptedEi := adaptToKeyValue(ei)
+		newValues := map[string]dosa.FieldValue{
+			key:   cacheKey,
+			value: cacheValue,
+		}
+		return c.fallback.Upsert(ctx, adaptedEi, newValues)
```

Comment:
I think calls to the fallback should extend the parent context to prevent spurious timeouts caused by high latency in the origin's response.